src/linker/Linker/DynamicDependency.cs

```diff
@@ -0,0 +1,143 @@
+		{
+			MemberSignature = memberSignature;
+			TypeName = typeName;
+			AssemblyName = assemblyName;
+		}
+
+		public DynamicDependency (DynamicallyAccessedMemberTypes memberTypes, TypeReference type)
+		{
+			MemberTypes = memberTypes;
+			Type = type;
+		}
+
+		public DynamicDependency (DynamicallyAccessedMemberTypes memberTypes, string typeName, string assemblyName)
+		{
+			MemberTypes = memberTypes;
+			TypeName = typeName;
+			AssemblyName = assemblyName;
+		}
+
+		public string? MemberSignature { get; }
+
+		public DynamicallyAccessedMemberTypes MemberTypes { get; }
+
+		public TypeReference? Type { get; }
+
+		public string? TypeName { get; }
+
+		public string? AssemblyName { get; }
+
+		public string? Condition { get; set; }
+
+		public static DynamicDependency? ProcessAttribute (LinkContext context, ICustomAttributeProvider provider, CustomAttribute customAttribute)
+		{
+			if (!(provider is MethodDefinition || provider is FieldDefinition))
+				return null;
+			var member = provider as IMemberDefinition;
+
+			if (!ShouldProcess (context, customAttribute))
+				return null;
+
+			var dynamicDependency = GetDynamicDependency (context, customAttribute);
+			if (dynamicDependency != null)
+				return dynamicDependency;
+
+			context.LogMessage (MessageContainer.CreateWarningMessage (context,
+				$"Invalid DynamicDependencyAttribute on '{member}'",
+				2030, MessageOrigin.TryGetOrigin (member)));
+			return null;
+		}
+
+		static DynamicDependency? GetDynamicDependency (LinkContext context, CustomAttribute ca)
+		{
+			var args = ca.ConstructorArguments;
+			if (args.Count < 1 || args.Count > 3)
+				return null;
+
+			// First argument is string or DynamicallyAccessedMemberTypes
+			string? memberSignature = args[0].Value as string;
+			if (args.Count == 1)
+				return memberSignature == null ? null : new DynamicDependency (memberSignature);
+			DynamicallyAccessedMemberTypes? memberTypes = null;
+			if (memberSignature == null) {
+				var argType = args[0].Type;
+				if (!argType.IsTypeOf<DynamicallyAccessedMemberTypes> ())
+					return null;
+				try {
+					memberTypes = (DynamicallyAccessedMemberTypes) args[0].Value;
+				} catch (InvalidCastException) { }
+				if (memberTypes == null)
+					return null;
+			}
+
+			// Second argument is Type for ctors with two args, string for ctors with three args
+			if (args.Count == 2) {
+				if (!(args[1].Value is TypeReference type))
+					return null;
+				return memberSignature == null ? new DynamicDependency (memberTypes!.Value, type) : new DynamicDependency (memberSignature, type);
+			}
+			Debug.Assert (args.Count == 3);
+			if (!(args[1].Value is string typeName))
+				return null;
+
+			// Third argument is assembly name
+			if (!(args[2].Value is string assemblyName))
+				return null;
+
+			var dynamicDependency = memberSignature == null ? new DynamicDependency (memberTypes!.Value, typeName, assemblyName) : new DynamicDependency (memberSignature, typeName, assemblyName);
```

Comment:
```suggestion
			return new DynamicDependency (memberSignature ?? memberTypes!.Value, typeName, assemblyName) {
				OriginalAttribute = ca
			};
```

Reply:
That doesn't type-check because `memberTypes!.Value` isn't a string.

Reply:
What type check?

Reply:
It needs to call a different ctor for string vs DynamicallyAccessedMemberTypes. `memberSignature ?? memberTypes!.Value` doesn't have a well-defined type - it produces `error CS0019: Operator '??' cannot be applied to operands of type 'string' and 'DynamicallyAccessedMemberTypes'`.